src/AlgebraicGeometry/TropicalGeometry/groebner_fan.jl

```diff
@@ -0,0 +1,541 @@
+ 1
+ 1
+
+julia> G2 = [x1^2+x2,x2^2+x3,x3]
+3-element Vector{QQMPolyRingElem}:
+ x1^2 + x2
+ x2^2 + x3
+ x3
+
+julia> homogeneity_vector(G2)
+3-element Vector{ZZRingElem}:
+ 1
+ 2
+ 4
+
+julia> G3 = [x1^2+x2,x2^2+x3,x3+1]
+3-element Vector{QQMPolyRingElem}:
+ x1^2 + x2
+ x2^2 + x3
+ x3 + 1
+
+julia> homogeneity_vector(G3) == nothing
+true
+
+```
+"""
+function homogeneity_vector(G::Vector{<:MPolyElem})
+    homogeneitySpace = homogeneity_space(G)
+
+    # return nothing if homogeneitySpace is the origin
+    # (which will often be the case)
+    if dim(homogeneitySpace)<1
+        return nothing
+    end
+
+    # test if homogeneitySpace contains the ones vector
+    n = ambient_dim(homogeneitySpace)
+    homogeneityVector = ones(fmpz,n)
+    if homogeneityVector in homogeneitySpace
+        return homogeneityVector
+    end
+
+    # try finding any positive vector inside of it
+    positiveOrthant = positive_hull(identity_matrix(ZZ,n))
+    posHomogeneitySpace = intersect(homogeneitySpace,positiveOrthant)
+    homogeneityVector = unique_identifying_point(posHomogeneitySpace)
+    if isnothing(findfirst(vi -> (vi<=0), homogeneityVector))
+        return homogeneityVector
+    end
+
+    # return nothing if everything fails
+    return nothing
+end
+
+
+
+@doc raw"""
+    maximal_groebner_cone(G::Oscar.IdealGens{<:MPolyElem}; homogeneityWeight::Union{Nothing,Vector{fmpz}}=nothing)
+
+Returns the maximal Groebner cone of a Groebner basis `G`, i.e., the closure of all weight vectors with respect to whose weighted ordering `G` is a Groebner basis (independent of tie-breaker).
+
+If `homogeneityWeight==nothing`, assumes that `G` is not quasi-homogeneous, i.e. not homogeneous with respect to any positive weight vector, and returns a cone inside the positive orthant.
+
+If `homogeneityWeight!=nothing`, assumes that `G` is quasi-homogeneous with respect to it and returns a cone whose lineality space contains `homogeneityWeight`.
+
+If `homogeneityWeight` is unspecified, will test if `G` is quasi-homogeneous, and behave accordingly.
+
+# Examples
+```jldoctest
+julia> Qx,(x1,x2,x3) = PolynomialRing(QQ,["x1","x2","x3"])
+(Multivariate Polynomial Ring in x1, x2, x3 over Rational Field, QQMPolyRingElem[x1, x2, x3])
+
+julia> I = ideal([x1,x2+x3])
+ideal(x1, x2 + x3)
+
+julia> G = groebner_basis(I,ordering=lex(Qx))
+Gröbner basis with elements
+1 -> x2 + x3
+2 -> x1
+with respect to the ordering
+lex([x1, x2, x3])
+
+julia> rays_modulo_lineality(maximal_groebner_cone(G))
+(rays_modulo_lineality = RayVector{QQFieldElem}[[0, 0, -1]], lineality_basis = RayVector{QQFieldElem}[[1, 0, 0], [0, 1, 1]])
+
+```
+"""
+function maximal_groebner_cone(G::Oscar.IdealGens{<:MPolyElem})
+    @assert is_groebner_basis(G)
+    ord = ordering(G)
+    G = collect(G)
+    homogeneityWeight = homogeneity_vector(G)
+    return maximal_groebner_cone(G,ord,homogeneityWeight)
+end
+function maximal_groebner_cone(G::Oscar.IdealGens{<:MPolyElem}; homogeneityWeight::Vector{fmpz}=fmpz[])
```

Comment:
This is a minor nitpick, and not a merge blocker, but: please put an empty line between the end of one function and the start of the next
```suggestion

function maximal_groebner_cone(G::Oscar.IdealGens{<:MPolyElem}; homogeneityWeight::Vector{fmpz}=fmpz[])
```